src/crystal/system/unix/kqueue/event_loop.cr

```diff
@@ -0,0 +1,230 @@
+      elsif event = pd.value.@readers.ready!
+        resume_io(event)
+      end
+    when LibC::EVFILT_WRITE
+      if (kevent.value.fflags & LibC::EV_ERROR) == LibC::EV_ERROR
+        # OPTIMIZE: pass errno (kevent.data) through PollDescriptor
+        pd.value.@writers.consume_each { |event| resume_io(event) }
+      elsif event = pd.value.@writers.ready!
+        resume_io(event)
+      end
+    end
+  end
+
+  def interrupt : Nil
+    return unless @interrupted.test_and_set
+
+    {% if LibC.has_constant?(:EVFILT_USER) %}
+      @kqueue.kevent(
+        INTERRUPT_IDENTIFIER,
+        LibC::EVFILT_USER,
+        LibC::EV_ADD | LibC::EV_ONESHOT,
+        LibC::NOTE_FFCOPY | LibC::NOTE_TRIGGER | 1_u16)
+    {% else %}
+      byte = 1_u8
+      ret = LibC.write(@pipe[1], pointerof(byte), sizeof(typeof(byte)))
+      raise RuntimeError.from_errno("write") if ret == -1
+    {% end %}
+  end
+
+  protected def system_add(fd : Int32, gen_index : Int64) : Nil
+    Crystal.trace :evloop, "kevent", op: "add", fd: fd, gen_index: gen_index
+
+    # register both read and write events
+    kevents = uninitialized LibC::Kevent[2]
+    2.times do |i|
+      kevent = kevents.to_unsafe + i
+      filter = i == 0 ? LibC::EVFILT_READ : LibC::EVFILT_WRITE
+      System::Kqueue.set(kevent, fd, filter, LibC::EV_ADD | LibC::EV_CLEAR, udata: Pointer(Void).new(gen_index.to_u64!))
+    end
+
+    @kqueue.kevent(kevents.to_slice) do
+      # we broadly add file descriptors to kqueue whenever we open them, but
+      # sometimes the other end is closed and registration can fail (e.g.
+      # stdio).
+      #
+      # we can safely discard these errors since further read or write to these
+      # file descriptors will fail with the same error and the evloop will never
+      # try to wait.
+      unless Errno.value.in?(Errno::ENODEV, Errno::EPIPE)
+        raise RuntimeError.from_errno("kevent")
+      end
+    end
+  end
+
+  protected def system_del(fd : Int32) : Nil
+    # nothing to do: close(2) will do the job
+  end
+
+  protected def system_del(fd : Int32, &) : Nil
+    # nothing to do: close(2) will do the job
+  end
+
+  private def system_set_timer(time : Time::Span?) : Nil
+    if time
+      flags = LibC::EV_ADD | LibC::EV_ONESHOT | LibC::EV_CLEAR
+      t = time - Time.monotonic
+      data =
+        {% if LibC.has_constant?(:NOTE_NSECONDS) %}
+          t.total_nanoseconds.to_i64!.clamp(0..)
+        {% else %}
+          # legacy BSD (and DragonFly) only have millisecond precision
+          t.positive? ? t.total_milliseconds.to_i64!.clamp(1..) : 0
+        {% end %}
+    else
+      flags = LibC::EV_DELETE
+      data = 0_u64
+    end
+
+    fflags =
+      {% if LibC.has_constant?(:NOTE_NSECONDS) %}
+        LibC::NOTE_NSECONDS
+      {% else %}
+        0
+      {% end %}
+
+    # use the evloop address as the unique identifier for the timer kevent
+    ident = LibC::SizeT.new!(self.as(Void*).address)
```

Comment:
~~We'd still need to cast to `SizeT` (actually it's `UintptrT` but we don't have it) because `object_id` always returns an `UInt64`:~~

```crystal
ident = LibC::SizeT.new!(object_id)
```

Scratch that: I can just use an arbitrary number, as I did for `INTERRUPT_IDENTIFIER`.